docs/spec/lightclient/detection/req-ibc-detection.md

```diff
@@ -0,0 +1,296 @@
+  "evidence" that there was a fork. 
+  
+- The relayer can read the handler's consensus state. Thus the relayer can
+  feed the handler precisely the information the handler needs to detect a
+  fork. 
+  What is this
+  information needs to be specified.
+  
+- The information depends on the verification the handler does. It
+  might be necessary to provide a bisection proof (list of
+  lightblocks) so that the handler can verify based on its local
+  lightstore a header *h* that is conflicting with a header *h'* in the
+  local lightstore, that is, *h != h'* and *h.Height = h'.Height*
+  
+### Relayer requirement: Fork detection
+
+Let's assume there is a fork at chain A. There are two ways the
+relayer can figure that out:
+
+1. as the relayer contains a light client for A, it also includes a fork
+   detector that can detect a fork.
+	  
+2. the relayer may also detect a fork by observing that the
+   handler for chain A (on chain B)
+   is on a different branch than the relayer
+
+
+- in both detection scenarios, the relayer should submit evidence to 
+  full nodes of chain A where there is a fork. As we assume a fullnode
+  has a complete list of blocks, it is sufficient to send "Bucky's
+  evidence" (https://github.com/tendermint/tendermint/issues/5083),
+  that is, 
+     - two lightblocks from different branches + 
+	 - a lightblock (perhaps just a height) from which both blocks 
+	   can be verified.
+  
+- in the scenario 2., the relayer must feed the A-handler (on chain B)
+  a proof of a fork on A so that chain B can react accordingly
+  
+### Handler requirement
+  
+- there are potentially many relayers, some correct some faulty
+
+- a handler cannot trust the information provided by the relayer, 
+  but must verify 
+  (Доверя́й, но проверя́й)
+
+- in case of a fork, we accept that the handler temporarily stores
+  headers (tagged as verified). 
+  
+- eventually, a handler should be informed
+ (`checkMisbehaviourAndUpdateState`)
+ by some relayer that it has
+  verified a header from a fork. Then the handler should do what is
+ required by IBC in this case (stop?)
+
+### Challenges in the handler requirement
+
+- handlers and relayers work on different lightstores. In principle
+  the lightstore need not intersect in any heights a priori
+
+- if a relayer  sees a header *h* it doesn't know at a handler (`queryChainConsensusState`), the
+  relayer needs to
+  verify that header. If it cannot do it locally based on downloaded
+  and verified (trusted?) light blocks, it might need to use
+  `VerifyToTarget` (bisection). To call `VerifyToTarget` we might keep
+  *h* in the lightstore. If verification fails, we need to download the
+  "alternative" header of height *h.Height* to generate evidence for
+  the handler.
+  
+- we have to specify what precisely `queryChainConsensusState`
+  returns. It cannot be the complete lightstore. Is the last header enough?
+
+- we would like to assume that every now and then (smaller than the
```

Comment:
Do we need the relayer to check *every single* ConsensusState? Is it possible that the handler could be tricked onto a fork for just one height and then brought back to the main chain in such a way that if relayers don't check every ConsensusState inserted by the handler, they could actually miss this?

Reply:
This is not so clear for me. What I understood from recent discussions is that the application state stores the history -- at least several (hundred) blocks back. If this is true, I believe one cannot go back from a branch without violating invariants in the application state that should show up in some checks. But this needs to be confirmed.